From e46af709e65ab0e954090f275fe03ef434d2c2f6 Mon Sep 17 00:00:00 2001
From: Pankaj Dev <pankaj.dev@st.com>
Date: Fri, 9 May 2025 09:12:29 +0530
Subject: [PATCH 829/841] fix(st-usb): issue stall on EP0 inside usb-stop

In case some Control transaction is incomplete while usb-stop is
issued then the Ctrl-EP state machine needs to be reset,
otherwise creates bad Ctrl-state when usb-start is issued later

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 1760fbe31dc5eda22f4300e3ad16f327534e115d fix(st-usb): issue stall on EP0 inside usb-stop]
Signed-off-by: Pankaj Dev <pankaj.dev@st.com>
Change-Id: Ie3e408246e3d9b85cc89644109bee0140f60594b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/457600
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Tested-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 drivers/st/usb_dwc3/usb_dwc3.c | 28 ++++++++++++++++++----------
 1 file changed, 18 insertions(+), 10 deletions(-)

diff --git a/drivers/st/usb_dwc3/usb_dwc3.c b/drivers/st/usb_dwc3/usb_dwc3.c
index 1a4aaca09..a0920a674 100644
--- a/drivers/st/usb_dwc3/usb_dwc3.c
+++ b/drivers/st/usb_dwc3/usb_dwc3.c
@@ -940,12 +940,24 @@ static enum usb_status dwc3_ep_stop_xfer(dwc3_handle_t *dwc3_handle, struct usbd
 	return ret;
 }
 
+static enum usb_status dwc3_ep_set_stall(dwc3_handle_t *dwc3_handle,
+					 const usb_dwc3_endpoint_t *dwc3_ep)
+{
+	dwc3_epcmd_params_t params;
+
+	(void)memset(&params, 0x00, sizeof(params));
+
+	return dwc3_execute_dep_cmd(dwc3_handle, dwc3_ep->phy_epnum, USB_DWC3_DEPCMD_SETSTALL,
+				    &params);
+}
+
 static enum usb_status usb_dwc3_stop_device(void *handle)
 {
 	dwc3_handle_t *dwc3_handle = (dwc3_handle_t *)handle;
 	uint64_t timeout;
 	uint8_t i;
 	uint32_t evtcnt;
+	enum usb_status ret;
 
 	/*
 	 * Stop transfers for all(USB_DWC3_NUM_IN_EPS) EP
@@ -960,6 +972,12 @@ static enum usb_status usb_dwc3_stop_device(void *handle)
 		dwc3_ep_stop_xfer(dwc3_handle, &dwc3_handle->pcd_handle->out_ep[i]);
 	}
 
+	/* Issue SetStall on EP0 to reset Ctrl-EP state machine */
+	ret = dwc3_ep_set_stall(dwc3_handle, &dwc3_handle->OUT_ep[0]);
+	if (ret != USBD_OK) {
+		ERROR("%s: EP0 stall failed %u\n", __func__, ret);
+	}
+
 	/*
 	 * In the Synopsis DesignWare Cores USB3 Databook Rev. 3.30a
 	 * Section 1.3.4, it mentions that for the DEVCTRLHLT bit, the
@@ -1018,16 +1036,6 @@ static enum usb_status usb_dwc3_ep0_start_xfer(void *handle, struct usbd_ep *ep)
 	return usb_dwc3_ep_start_xfer(handle, ep);
 }
 
-static enum usb_status dwc3_ep_set_stall(dwc3_handle_t *dwc3_handle, usb_dwc3_endpoint_t *dwc3_ep)
-{
-	dwc3_epcmd_params_t params;
-
-	(void)memset(&params, 0x00, sizeof(params));
-
-	return dwc3_execute_dep_cmd(dwc3_handle, dwc3_ep->phy_epnum, USB_DWC3_DEPCMD_SETSTALL,
-				    &params);
-}
-
 static enum usb_status usb_dwc3_ep_set_stall(void *handle, struct usbd_ep *ep)
 {
 	dwc3_handle_t *dwc3_handle = (dwc3_handle_t *)handle;
-- 
2.47.2

